File: src/main/java/botmanager/frostbalance/commands/influence/SupportCommand.java
package botmanager.frostbalance.commands.influence;

import botmanager.Utilities;
import botmanager.frostbalance.Frostbalance;
import botmanager.frostbalance.Influence;
import botmanager.frostbalance.MemberWrapper;
import botmanager.frostbalance.UserWrapper;
import botmanager.frostbalance.command.AuthorityLevel;
import botmanager.frostbalance.command.FrostbalanceHybridCommandBase;
import botmanager.frostbalance.command.GenericMessageReceivedEventWrapper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;

/**
 *
 * @author MC_2018 <mc2018.git@gmail.com>
 */
public class SupportCommand extends FrostbalanceHybridCommandBase {

    private static final double PRIVATE_MODIFIER = 0.5;

    public SupportCommand(Frostbalance bot) {
        super(bot, new String[] {
                bot.getPrefix() + "support",
                bot.getPrefix() + "s"
        }, AuthorityLevel.GENERIC, Condition.GUILD_EXISTS);
    }

    @Override
    protected void runHybrid(GenericMessageReceivedEventWrapper eventWrapper, String... params) {
        List<String> resultLines = new ArrayList<>();

        Influence transferAmount = new Influence(params[params.length - 1]);
        MemberWrapper bMember = eventWrapper.getBotMember().get();

        String targetName = String.join(" ", Arrays.copyOfRange(params, 0, params.length - 1));
        Optional<UserWrapper> targetUser = bot.getUserByName(targetName);
        if (!targetUser.isPresent()) {
            resultLines.add("Could not find user '" + targetName + "'.");
            eventWrapper.sendResponse(resultLines);
            return;
        }
        MemberWrapper targetMember = targetUser.get().getMember(eventWrapper.getGuildId().get());

        if (transferAmount.greaterThan(bMember.getInfluence())) {
            transferAmount = bMember.getInfluence();
            resultLines.add("You don't have that much influence to give. You will instead use all of your influence.");
        } else if (transferAmount.isNegative() || !transferAmount.isNonZero()) { //'else' allows you to bluff when you have 0 influence.
            resultLines.add("You have to spend *some* influence to support someone.");
            eventWrapper.sendResponse(resultLines);
            return;
        }

        if (targetMember.equals(bMember)) {
            resultLines.add("You give yourself " + transferAmount + " influence in " + eventWrapper.getGuild().get().getName() + " because you are awesome.");
            eventWrapper.sendResponse(resultLines);
            return;
        }

        bMember.adjustInfluence(transferAmount.negate());

        if (eventWrapper.isPublic()) {
            eventWrapper.getMessage().delete().queue();
            resultLines.add(bMember.getEffectiveName() + " has *supported* " + targetMember.getEffectiveName() + ", increasing their influence here.");
            Utilities.sendPrivateMessage(targetMember.getUserWrapper().getUser().get(),
                    String.format("%s has *supported* you, increasing your influence in %s by %s.",
                            bMember.getEffectiveName(),
                            eventWrapper.getBotGuild().get().getName(),
                            transferAmount));
            targetMember.adjustInfluence(transferAmount);
        } else {
            resultLines.add("You have *supported* " + targetMember.getEffectiveName() + " secretly, increasing their influence in " + eventWrapper.getBotGuild().get().getName() + ".");
            Utilities.sendPrivateMessage(targetMember.getUserWrapper().getUser().get(),
                    String.format("You have been supported secretly by " + bMember.getEffectiveName() + ". Your influence in %s has been increased by %s.",
                            eventWrapper.getBotGuild().get().getName(),
                            transferAmount.applyModifier(PRIVATE_MODIFIER)));
            targetMember.adjustInfluence(transferAmount.applyModifier(PRIVATE_MODIFIER));
        }
        eventWrapper.sendResponse(resultLines);
        return;

    }

    @Override
    protected String info(AuthorityLevel authorityLevel, boolean isPublic) {
        if (isPublic) {
            return "**" + bot.getPrefix() + "__s__upport PLAYER AMOUNT** - Support another player, giving them the set amount of influence";
        } else {
            return "**" + bot.getPrefix() + "__s__upport PLAYER AMOUNT** - Support another player privately (they know who you are), giving them 50% of what you give choose to give";
        }
    }

}

File: src/main/java/botmanager/frostbalance/commands/influence/SupportCommand.kt
package botmanager.frostbalance.commands.influence

import botmanager.Utilities
import botmanager.frostbalance.Frostbalance
import botmanager.frostbalance.command.*
import botmanager.frostbalance.menu.ConfirmationMenu
import java.lang.NumberFormatException
import java.util.*

class SupportCommand(bot: Frostbalance) : FrostbalanceGuildCommand(bot, arrayOf(
        "support",
        "s"
), AuthorityLevel.GENERIC, ContextLevel.ANY) {

    override fun executeWithGuild(context: GuildMessageContext, vararg params: String) {
        val arguments = ArgumentStream(params)

        val resultLines: MutableList<String> = ArrayList()
        val bMember = context.member!!
        val targetName = arguments.exhaust(1)
        val targetMember = bot.getUserByName(targetName, context.guild)?.memberIfIn(context.guild)
        if (targetMember == null) {
            resultLines.add("Could not find user '$targetName'.")
            context.sendMultiLineResponse(resultLines)
            return
        }
        var transferAmount = try {
            arguments.nextInfluence() ?: return context.sendResponse("No influence found!")
        } catch (e: NumberFormatException) {
            return context.sendResponse("That last bit wasn't a number. Try again.")
        }
        if (transferAmount.greaterThan(bMember.influence)) {
            transferAmount = bMember.influence
            resultLines.add("You don't have that much influence to give. You will instead use all of your influence.")
        } else if (transferAmount.isNegative || !transferAmount.nonZero) { //'else' allows you to bluff when you have 0 influence.
            resultLines.add("You have to spend *some* influence to support someone.")
            context.sendMultiLineResponse(resultLines)
            return
        }
        if (targetMember == bMember) {
            resultLines.add("You support yourself with " + transferAmount + " influence in ${context.guild.name}. You should return the favor.")
            context.sendMultiLineResponse(resultLines)
            return
        }
        bMember.adjustInfluence(transferAmount.negate())
        if (context.isPublic) {
            context.message.delete().queue()
            resultLines.add(bMember.effectiveName + " has *supported* " + targetMember.effectiveName + ", increasing their influence here.")
            Utilities.sendPrivateMessage(targetMember.userWrapper.jdaUser, context.buildEmbed(String.format("%s has *supported* you, increasing your influence in %s by %s.",
                    bMember.effectiveName,
                    context.guild!!.name,
                    transferAmount), false))
            targetMember.adjustInfluence(transferAmount)
            context.sendMultiLineResponse(resultLines)
        } else {
            ConfirmationMenu(bot, context, {
                if (transferAmount.applyModifier(PRIVATE_MODIFIER) > 0) {
                    resultLines.add("You have *supported* " + targetMember.effectiveName + " secretly, increasing their influence in " + context.guild.name + " by ${transferAmount.applyModifier(PRIVATE_MODIFIER)}.")
                    Utilities.sendPrivateMessage(targetMember.userWrapper.jdaUser, context.buildEmbed(String.format("You have been supported secretly by " + bMember.effectiveName + ". Your influence in %s has been increased by %s.",
                            context.guild.name,
                            transferAmount.applyModifier(PRIVATE_MODIFIER)), false))
                    targetMember.adjustInfluence(transferAmount.applyModifier(PRIVATE_MODIFIER))

                    val refundAmount = transferAmount.remainderOfModifier(PRIVATE_MODIFIER)
                    if (refundAmount.nonZero) {
                        resultLines.add("You have been refunded $refundAmount that would have gone unused.")
                        bMember.adjustInfluence(transferAmount)
                    }
                } else {
                    resultLines.add("After rounding, your support would have no effect. Your influence has been refunded.")
                    bMember.adjustInfluence(transferAmount)
                }
                context.sendMultiLineResponse(resultLines)
            }, "Are you sure you want to support ${targetMember.effectiveName} privately? Only ${PRIVATE_MODIFIER*100}% of your influence will be used.")
                    .send(context.channel, context.author)
        }
        return
    }

    override fun info(authorityLevel: AuthorityLevel?, isPublic: Boolean): String? {
        return if (isPublic) {
            "**" + bot.prefix + "__s__upport PLAYER AMOUNT** - Support another player, giving them the set amount of influence"
        } else {
            "**" + bot.prefix + "__s__upport PLAYER AMOUNT** - Support another player privately (they know who you are), giving them 50% of what you give choose to give"
        }
    }

    companion object {
        private const val PRIVATE_MODIFIER = 0.5
    }
}